crates/proof-of-sql/src/base/database/union_util.rs

```diff
@@ -0,0 +1,337 @@
+use super::{
+    Column, ColumnField, ColumnOperationError, ColumnOperationResult, ColumnType, Table,
+    TableOperationError, TableOperationResult, TableOptions,
+};
+use crate::base::scalar::Scalar;
+use alloc::vec::Vec;
+use bumpalo::Bump;
+
+/// Check if two schemas are compatible
+/// Note that we can tolerate differences in column names but not in column types
+fn are_schemas_compatible(left: &[ColumnField], right: &[ColumnField]) -> bool {
+    left.len() == right.len()
+        && left
+            .iter()
+            .zip(right)
+            .all(|(field1, field2)| field1.data_type() == field2.data_type())
+}
+
+/// Union multiple columns of the same type into a single column
+///
+/// # Panics
+/// This function should never panic as long as it is written correctly
+pub fn column_union<'a, S: Scalar>(
+    columns: &[&Column<'a, S>],
+    alloc: &'a Bump,
+    column_type: ColumnType,
+) -> ColumnOperationResult<Column<'a, S>> {
+    // Check for type mismatch
+    let possible_bad_column_type = columns.iter().find_map(|col| {
+        let found_column_type = col.column_type();
+        (found_column_type != column_type).then_some(found_column_type)
+    });
+    if let Some(bad_column_type) = possible_bad_column_type {
+        return Err(ColumnOperationError::UnionDifferentTypes {
+            actual_type: bad_column_type,
+            correct_type: column_type,
+        });
+    }
+    Ok(match column_type {
+        ColumnType::Boolean => {
+            let result: Vec<_> = columns
+                .iter()
+                .flat_map(|col| col.as_boolean().expect("Column types should match"))
+                .copied()
+                .collect();
```

Comment:
Should be no need to collect here.
Use `alloc_slice_fill_iter` or look at the source of `alloc_slice_fill_iter` (i.e. use `alloc_slice_fill_with`) instead.